pubtools/_quay/security_manifest_pusher.py

```diff
@@ -0,0 +1,535 @@
+                LOG.info(f"Existing attestation file found for image {image_ref}")
+                existing_security_manifest = self.get_security_manifest_from_attestation(
+                    attestation_file
+                )
+                existing_products = self.security_manifest_get_products(existing_security_manifest)
+
+                # It is assumed that images with the same digest must have the same
+                # content of security manifests
+                if len(products - existing_products) == 0:
+                    LOG.info(
+                        "No new product will be added to the existing security manifest. Skipping "
+                        f"the security manifest upload for image {image_ref}."
+                    )
+                    continue
+                LOG.info(
+                    f"Image {image_ref} already has an attestation. A new attestation will "
+                    "be uploaded containing all product names."
+                )
+
+                self.delete_existing_attestation(image_ref, dir_path)
+                products = products | existing_products
+
+            if products:
+                full_security_manifest_path = self.security_manifest_add_products(
+                    image_manifest.security_manifest_path, products
+                )
+            else:
+                LOG.warning(
+                    f"Push item {item} doesn't contain a product name. A new attestation "
+                    "will be created without this information."
+                )
+                full_security_manifest_path = image_manifest.security_manifest_path
+
+            self.cosign_attest_security_manifest(full_security_manifest_path, image_ref)
+
+    def get_source_item_security_manifest(
+        self, item: object, dir_path: str
+    ) -> list[DigestSecurityManifest]:
+        """
+        Get security manifest of a source image.
+
+        Source image has only one architecture, so at most one security manifest will be present.
+        Based on testing, source images don't seem to contain security manifests, so failure to
+        extract it doesn't result in an error.
+
+        Args:
+            item (ContainerPushItem):
+                Push item of a source image.
+            dir_path (str):
+                Directory path where the manifest file should be created.
+        Returns [DigestSecurityManifest]:
+            Extracted security manifest.
+        """
+        source_ref = item.metadata["pull_url"]
+        digest = self.src_quay_client.get_manifest_digest(source_ref)
+        security_manifest_path = os.path.join(
+            dir_path, f"security_manifest_source_{uuid.uuid4().hex}.json"
+        )
+
+        result = self.cosign_get_security_manifest(source_ref, security_manifest_path)
+        if not result:
+            LOG.warning(f"Source image {source_ref} doesn't contain a security manifest")
+            return []
+
+        return [
```

Comment:
Why list when there's always only one item in it?

Reply:
Multiarch method produces multiple entries. These results are iterated in the push_item_security_manifests method, so a list is expected regardless of how many entries it contains.

Reply:
Yeah, I understand that. But if method does always return just one item in the list it doesn't make sense to return the list. You should wrap result of this method in the list when this method is called. Method written like this is confusing.

Reply:
There are two similair methods: get_source_item_security_manifest and get_multiarch_item_security_manifests. If pubtools-quay starts supporting another push item type, a new similiar method will be added. To me, it makes most sense that inputs and outputs of these methods follow the same pattern, and can be worked with in the same manner regardless of which method is called.

Reply:
Can you the at least rename get_source_item_security_manifest to **manifests** to at least make clear method returns list

Reply:
Changed.